system/jlib/jlzw.cpp

```diff
@@ -2229,6 +2252,142 @@ class CCompressedFile : implements ICompressedFileIO, public CInterface
+            assertex(r==toread);
+        }
+        else
+        {
+            assertex((memsize_t)trailer.indexPos==trailer.indexPos);
+            memcpy(indexbuf.reserveTruncate(toread),mmfile->base()+(memsize_t)trailer.indexPos,toread);
+        }
+
+        //Ensure that an expander is set up
+        if (trailer.recordSize==0)
+        {
+            if (!expander)
+            {
+                if (compMethod == COMPRESS_METHOD_FASTLZ)
+                    expander.setown(createFastLZExpander());
+                else if (compMethod == COMPRESS_METHOD_LZ4)
+                    expander.setown(createLZ4Expander());
+                else // fallback
+                {
+                    compMethod = COMPRESS_METHOD_LZW;
+                    expander.setown(createLZWExpander(true));
+                }
+            }
+        }
+    }
+
+    virtual ~CCompressedFileReader()
+    {
+        try { close(); }
+        catch (IException *e)
+        {
+            EXCLOG(e, "~CCompressedFile");
+            e->Release();
+        }
+    }
+
+    virtual size32_t read(offset_t pos, size32_t len, void * data) override
+    {
+        CriticalBlock block(crit);
+        size32_t ret=0;
+        while (pos<trailer.expandedSize) {
+            if ((offset_t)len>trailer.expandedSize-pos)
+                len = (size32_t)(trailer.expandedSize-pos);
+            if ((pos>=curblockpos)&&(pos<curblockpos+iobuffer.length())) { // see if in current buffer
+                size32_t tocopy = (size32_t)(curblockpos+iobuffer.length()-pos);
+                if (tocopy>len)
+                    tocopy = len;
+                memcpy(data,iobuffer.toByteArray()+(pos-curblockpos),tocopy);
+                ret += tocopy;
+                len -= tocopy;
+                data = (byte *)data+tocopy;
+                pos += tocopy;
+            }
+            if (len==0)
+                break;
+            getblock(pos);
+        }
+        return ret;
+    }
+
+    virtual offset_t size() override
+    { 
+        CriticalBlock block(crit);
+        return trailer.expandedSize;
+    }
+
+    virtual size32_t write(offset_t pos, size32_t len, const void * data) override
+    {
+        throwUnexpected();
+    }
+    virtual void flush() override
+    {
+        throwUnexpected();
+    }
+    virtual void close() override
+    {
+        CriticalBlock block(crit);
+        if (fileio)
+            fileio->close();
+        curblockpos = 0;
+        curblocknum = (unsigned)-1; // relies on wrap
+    }
+
+// CCompressedFile impl.
+    virtual unsigned dataCRC() override
+    {
+        return trailer.datacrc;
+    }
+};
+
+//---------------------------------------------------------------------------------------------------------------------
+
+class CCompressedFileWriter : public CCompressedFileBase
+{
+    unsigned curblocknum;
+    offset_t curblockpos;           // logical pos (reading only)
+    MemoryBuffer iobuffer;          // buffer used for reading or writing
```

Comment:
Old comment from the original class. There are others here as well, following this one.